### PR TITLE
chore: remove now useless bls init

### PR DESCRIPTION
**Pull request by jeluard**
**Motivation**

Remove `bls` `init` code previously required for karma.

### CODE REVIEW

File: packages/light-client/src/index.ts [hunk 1]
```diff
@@ -149,15 +150,16 @@ export class Lightclient {
   static async initializeFromCheckpointRoot(
     args: Omit<LightclientInitArgs, "bootstrap"> & {
       checkpointRoot: phase0.Checkpoint["root"];
+      blsImplementation?: Implementation;
```
Comment (nflaig) [inline packages/light-client/src/index.ts]: Having an internal implementation as part of arguments seems bit strange to me, if this is really required can't we follow same approach used in ssz and discv5, i.e. using a [setter function](https://github.com/ChainSafe/discv5/blob/628876d353a3f444edaf29d78bfba8d0b27bb8f4/packages/enr/src/crypto.ts#L16)

another example
https://github.com/ChainSafe/lodestar/blob/918924eabad9ca615acf3d92ff50828fb40fedea/packages/cli/src/applyPreset.ts#L9

Comment (jeluard) [reply packages/light-client/src/index.ts]: `setHasher` is setting some global then? I think we should move away from those, they introduce fragility and break expectations.

e.g. `setImplementation` could be called after the init is done, making it void. 

There is a similar issue with the active preset, where you have to do things in the right order or it breaks. It also introduces coupling (or forces lack of coupling) between dependencies.

In general I fell like explicit is better over implicit (env variable, global variable), and letting user decide of implementation sounds like it.

Comment (nflaig) [reply packages/light-client/src/index.ts]: > they introduce fragility and break expectations.

How does this change expectations though? From the code I would expect now I can do this

```ts
lc1.initializeFromCheckpointRoot({blsImplementation: "herumi"})

// this light client instance will use different bls implementation
lc2.initializeFromCheckpointRoot({blsImplementation: "blst-native"})
```

but does this actually work and it that the intended use case?


> There is a similar issue with the active preset, where you have to do things in the right order or it breaks. It also introduces coupling (or forces lack of coupling) between dependencies.

This is a protection that ensures preset values can't be modified after those are loaded, much better to throw an error explicitly than to risk inconsistencies.

Comment (jeluard) [reply packages/light-client/src/index.ts]: > How does this change expectations though? From the code I would expect now I can do this

It looks marginally better to me but indeed there is still room to shoot yourself in the foot.

Your comment made me realize that having explicit `init` specifically here is probably wrong (and hopefully not required anymore as related to `karma`). Indeed, some code changes might force `init` to be already done before this code is reached.

A better solution might be:

* to remove`init` code from `light-client`
* to address configurability directly in `bls` library

Comment (nflaig) [reply packages/light-client/src/index.ts]: I see two patterns we can follow here
- either use dependency injection
- or load and set internally

If we wanna use dependency injection, the whole implementation should be passed imo and not just a name which then causing the library to be loaded internally.

I feel like the change in this PR was mixing both concepts.

> they introduce fragility and break expectations.

I think the safest approach is to allow setter to be called exactly once to ensure same implementation is used globally. Dependency injection gives more flexibility but also has the risk that you have to pass the implementation in possibly multiple places and you might accidentally miss one and it implicitly uses a default / fallback implementation.

That's how `setHasher` works and I think it's good for this use case as you would usually wanna use same hashing implementation across the whole project.

maybe @wemeetagain can give some more insights on this

Comment (jeluard) [reply packages/light-client/src/index.ts]: I shared some ideas here: https://github.com/ChainSafe/bls/issues/157

Comment (jeluard) [reply packages/light-client/src/index.ts]: With `setHasher`, user must ensure that the call is made at the right moment or there is a risk of having some default implementation being used (the one that kicks in when `setHasher` is not called).
Combined with other libs being intertwined, this looks a bit fragile to me (and hard to know for sure the call has been properly taken into consideration).

Comment (jeluard) [reply packages/light-client/src/index.ts]: Removed the implementation selection, this code was specific to karma anyway.

**Review by nflaig — APPROVED** [review]

**Comment by wemeetagain** [comment]
:tada: This PR is included in [v1.18.0](https://github.com/ChainSafe/lodestar/releases/tag/v1.18.0) :tada:

**Pull request merged**